clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2682,19 +2698,52 @@ void SYCLIntegrationHeader::emit(raw_ostream &O) {
   }
   O << "};\n\n";
 
+  O << "// array representing kernel parameters used in all kernels defined in "
+       "the\n";
+  O << "// corresponding source\n";
+  O << "static constexpr\n";
+  O << "const bool param_omit_table[] = {\n";
+  O << "  //OMIT_TABLE_BEGIN  \n";
+  for (unsigned I = 0; I < KernelDescs.size(); I++) {
+    auto &K = KernelDescs[I];
+    O << "  //--- " << K.Name << "\n";
+    O << "  ";
+    for (unsigned I = 0; I < K.Params.size(); I++) {
+      auto &P = K.Params[I];
+      for (unsigned J = 0; J < P.NumOpenCLParams; J++) {
+        O << "0"
+          << ", ";
+      }
+    }
+    O << "\n";
+  }
+  O << "  //OMIT_TABLE_END "
+    << "\n";
+  O << "  } \n\n";
+
   O << "// array representing signatures of all kernels defined in the\n";
   O << "// corresponding source\n";
   O << "static constexpr\n";
   O << "const kernel_param_desc_t kernel_signatures[] = {\n";
 
+  unsigned CurIndex = 0;
   for (unsigned I = 0; I < KernelDescs.size(); I++) {
     auto &K = KernelDescs[I];
     O << "  //--- " << K.Name << "\n";
-
     for (const auto &P : K.Params) {
       std::string TyStr = paramKind2Str(P.Kind);
       O << "  { kernel_param_kind_t::" << TyStr << ", ";
-      O << P.Info << ", " << P.Offset << " },\n";
+      O << P.Info << ", " << P.Offset << ", ";
+      for (unsigned X = 0; X < P.NumOpenCLParams; X++) {
```

Comment:
See if you can avoid the shift on the '0' element.  So this becomes something like:

if (X == 0)
  O << "param_omit_table["<< CurIndex << "]"
else
  O << " | (param_omit_table["<< CurIndex << "] << " << X << ")"

Reply:
Is the tabbing off in this function?  Also, see the above comment.

Reply:
Still this minor change.